portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java

```diff
@@ -0,0 +1,169 @@
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns a list of comments */
+@WebServlet("/list-comments")
+public class ListCommentsServlet extends HttpServlet {
+
+  // GET parameters
+  private String commentPicId;
+  private int fetchQuantity;
+  private boolean isOrderByDate;
+
+  @Override
+  public void init() {
+    fetchQuantity = 5;
+    isOrderByDate = true;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    handleGetParams(request);
+
+    // A pic id is needed to continue
+    if (commentPicId.isEmpty()) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Query fetchComments = buildQuery();
+
+    Iterable<Entity> commentsRetrieved = retrieveComments(fetchComments, fetchQuantity);
+
+    List<Comment> comments = createCommentsList(commentsRetrieved);
+
+    response.setContentType("application/json;");
+    response.getWriter().println(convertListToJson(comments));
+  }
+
+  /**
+   * Return the request parameter or the default value if the parameter
+   * was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null || value.isEmpty()) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  /**
+   * Get and assign parameters of a request
+   */
+  private void handleGetParams(HttpServletRequest request) {
+    String imageId = getParameter(request, "imageId", "");
+    if (imageId.isEmpty()) {
+      commentPicId = "";
+      return; // Return here since getting other parameters will be worthless
+    } else {
+      commentPicId = imageId;
+    }
+
+    String quantity = getParameter(request, "quantity", "");
+    if (quantity.isEmpty()) {
+      fetchQuantity = 5;
+    } else {
+      fetchQuantity = Integer.parseInt(quantity);
+    }
+
+    String dateOrder = getParameter(request, "dateOrder", "");
+    if (dateOrder.isEmpty()) {
+      isOrderByDate = false;
+    } else {
+      isOrderByDate = Boolean.parseBoolean(dateOrder);
+    }
+  }
+
+  /**
+   * Construct a query according to the parameters' values
+   */
+  private Query buildQuery() {
+    Query query = new Query("Comment");
+    
+    // Sort by date
+    if (isOrderByDate) {
```

Comment:
Perhaps a more descriptive member variable? They both are sorting by date in a way, just a matter of descending or ascending date. 

Also general comment about [ternary operators](https://en.wikipedia.org/wiki/%3F:#:~:text=In%20computer%20programming%2C%20%3F%3A%20is,true%2C%20and%20otherwise%20to%20c%20.), they are not necessary but a shorthand many people default to.

query.addSort("date", isOrderByDate ? SortDirection.DESCENDING : SortDirection.ASCENDING);

Reply:
Wanted to add my thoughts on this too. Ternary operators are definitely not required, but they are a cool to use if the "if" and "else" statements are almost identical.

Reply:
Okok, I'll think of another name. Cool to know that ternary operators are okay, I wasn't using them, because I wasn't sure if they would increase or decrease readability, I guess for this kind of statements they are cool.